pep-0458.txt

```diff
@@ -542,38 +542,49 @@ that disagrees with what is available on PyPI, which is indistinguishable from
-result, this proposal MUST address the problem of producing a consistent
-snapshot that captures the state of all known projects at a given time.  Each
-snapshot should safely coexist with any other snapshot, and be able to be
-deleted independently, without affecting any other snapshot.
+To keep TUF metadata on PyPI consistent with the highly volatile target files,
+consistent snapshots SHOULD be used. Each consistent snapshot captures the
+state of all known projects at a given time and MAY safely coexist with any
+other snapshot, or be deleted independently, without affecting any other
+snapshot.
+
+To maintain consistent snapshots, all TUF metadata MUST, when written to disk,
+include a version number in their filename:
 
-The solution presented in this PEP is that every metadata or data file managed
-by PyPI and written to disk MUST include in its filename the `cryptographic
-hash`__ of the file.  How would this help clients that use the TUF protocol to
-securely and consistently install or update a project from PyPI?
+  VERSION_NUMBER.ROLENAME.json,
+    where VERSION_NUMBER is an incrementing integer, and ROLENAME is one of the
+    top-level metadata roles -- *root*, *snapshot* or *targets* -- or one of
+    the delegated targets roles -- *bins* or *bin <i>*.
 
-__ https://en.wikipedia.org/wiki/Cryptographic_hash_function
+The only exception is the *timestamp* metadata file, whose version is not known
+in advance, when a client performs an update.
 
-The first step in the TUF protocol requires the client to download the latest
-*timestamp* metadata.  However, the client would not know in advance the hash
-of the *timestamp* associated with the latest snapshot.  Therefore, PyPI MUST
-redirect all HTTP GET requests for *timestamp* to the *timestamp* referenced in
-the latest snapshot.  The *timestamp* role is the root of a tree of
-cryptographic hashes that points to every other metadata that is meant to exist
-together (i.e., clients request metadata in timestamp -> snapshot -> root ->
-targets order).  Clients are able to retrieve any file from this snapshot
-by deterministically including, in the request for the file, the hash of the
-file in the filename.  Assuming infinite disk space and no `hash collisions`__,
-a client may safely read from one snapshot while PyPI produces another
+Instead the *timestamp* metadata serves as version root. That is, it lists the
+version of the *snapshot* metadata, which in turn lists the versions of *root*,
+*targets* and delegated targets metadata, all part of a given consistent
 snapshot.
 
+Eventually, *targets* or delegated targets metadata point to the actual target
+files, including their `cryptographic hashes`__. Thus, to mark a target file as
+part of a consistent snapshot it MUST, when written to disk, include its hash
+in its filename:
+
+  HASH.FILENAME
+    where HASH is the `hex digest`__ of the `SHA-256`__ hash of the file
```

Comment:
Do we only allow for sha-256 hashes?

Reply:
This was already in the pep, I just moved it up here. Should we leave the choice of algo open? It should be an algorithm that is also used when creating targets hashes, to be added in the (delegated) targets file. Because that's how the client builds the filename, right? (cc @JustinCappos, @trishankatdatadog).

Btw. the [spec](https://github.com/theupdateframework/specification/blob/master/tuf-spec.md#7-consistent-snapshots) does not prescribe an algorithm. It does, however, [point to an outdated version of this pep](https://github.com/theupdateframework/pep-on-pypi-with-tuf). We should remove those pointers in the spec.

Reply:
We can choose the secure hash algorithm here.  SHA-256 is fine.